### PR TITLE
Forbid types with required members from substituting for where T : new()

### PR DESCRIPTION
**Pull request by 333fred**
Specification: https://github.com/dotnet/csharplang/blob/main/proposals/required-members.md
Test plan: https://github.com/dotnet/roslyn/issues/57046

### CODE REVIEW

**Comment by 333fred** [comment]
@jcouv @RikkiGibson for review.

File: src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs [hunk 1]
```diff
@@ -1378,16 +1378,17 @@ private static bool HasPublicParameterlessConstructor(NamedTypeSymbol type, bool
...
             {
                 if (constructor.ParameterCount == 0)
                 {
-                    return constructor.DeclaredAccessibility == Accessibility.Public;
+                    return constructor.DeclaredAccessibility == Accessibility.Public
+                        && (!hasAnyRequiredMembers || constructor.HasSetsRequiredMembers);
                 }
             }
-            return synthesizedIfMissing;
+            return synthesizedIfMissing && !hasAnyRequiredMembers;
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs:1391]: It seems like we're making the assumption that the synthesized parameterless constructor will never have 'HasSetsRequiredMembers'. Is that fine?

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs (hunk 1 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs:1391]: I can't see how it wouldn't be.

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs [hunk 2]
```diff
@@ -4048,4 +4048,124 @@ void validate(ModuleSymbol module)
...
+
+    [Fact]
+    public void AllowRequiredAsNew_IndirectionViaStruct()
+    {
+        var code = """
+            M1<C>();
+
+            void M1<T>() where T : struct
+            {
+                M2<T>();
```
[the remark this answers is not in the document]
Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:4155]: No idea. I would love to, but no idea.

**Comment by jcouv** [comment]
>         private static bool HasPublicParameterlessConstructor(NamedTypeSymbol type, bool synthesizedIfMissing)

This method should be renamed (maybe "SatisfiesParameterlessConstructor"?) and doc comment updated.

---
Refers to: src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs:1377 in aac7cbe. [](commit_id = aac7cbe316673e8a6879097b5bac35fb7178f79e, deletion_comment = False)

File: src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs [hunk 3]
```diff
@@ -4048,4 +4048,124 @@ void validate(ModuleSymbol module)
...
+            {
+                public required int Prop1 { get; set; }
+            }
+            """;
+
+        var comp = CreateCompilationWithRequiredMembers(code);
+        comp.VerifyDiagnostics(
+            // (1,1): error CS0310: 'C' must be a non-abstract type with a public parameterless constructor in order to use it as parameter 'T' in the generic type or method 'M<T>()'
+            // M<C>();
+            Diagnostic(ErrorCode.ERR_NewConstraintNotSatisfied, "M<C>").WithArguments("M<T>()", "T", "C").WithLocation(1, 1)
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:4074]: Yes, that'd be more user-friendly

**Review by jcouv — COMMENTED** [review]
Done with review pass (iteration 1)

**Comment by 333fred** [comment]
@jcouv @RikkiGibson please take another look.

**Review by jcouv — APPROVED** [review]
LGTM Thanks (iteration 3)

**Review by RikkiGibson — APPROVED** [review]

File: src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs [hunk 4]
```diff
@@ -995,23 +997,34 @@ private static bool HasDuplicateInterfaces(NamedTypeSymbol type, ConsList<TypeSy
...
-                    Debug.Assert(error == ConstructorConstraintError.HasRequiredMembers);
-                    // '{2}' cannot satisfy the 'new()' constraint on parameter '{1}' in the generic type or or method '{0}' because '{2}' has required members.
-                    diagnosticsBuilder.Add(new TypeParameterDiagnosticInfo(typeParameter, new UseSiteInfo<AssemblySymbol>(new CSDiagnosticInfo(ErrorCode.ERR_NewConstraintCannotHaveRequiredMembers, containingSymbol.ConstructedFrom(), typeParameter, typeArgument.Type))));
-                }
                 return false;
             }
 
             return !hasError;
+
+            [MethodImpl(MethodImplOptions.NoInlining)]
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs:1007]: was this causing us to blow the stack in some test?

**Pull request merged**